File: basebuddy/static/js/integrations.js
(function () {
  const BOOL = [
    'MQTT_ENABLED', 'NOTIFY_ENABLED', 'NOTIFY_FALLBACK_GLOBAL', 'NOTIFY_INCLUDE_CLIP_DEFAULT',
    'MULTIPROC_DETECTION', 'LPR_ENABLED', 'SMTP_USE_TLS',
  ];
  const FIELDS = [
    'MQTT_HOST', 'MQTT_PORT', 'MQTT_TOPIC_PREFIX', 'MQTT_CLIENT_ID', 'MQTT_USERNAME', 'MQTT_PASSWORD',
    'NOTIFY_WEBHOOK_URL', 'NOTIFY_PUBLIC_BASE_URL',
    'TELEGRAM_BOT_TOKEN', 'TELEGRAM_CHAT_ID',
    'PUSHOVER_USER_KEY', 'PUSHOVER_API_TOKEN',
    'SMTP_HOST', 'SMTP_PORT', 'SMTP_USER', 'SMTP_PASSWORD', 'SMTP_FROM', 'SMTP_TO',
    'TWILIO_ACCOUNT_SID', 'TWILIO_AUTH_TOKEN', 'TWILIO_FROM_NUMBER', 'TWILIO_TO_NUMBER',
    'RECORDING_MODE', 'EVENT_CLIP_PRE_S', 'EVENT_CLIP_POST_S',
    'FFMPEG_HWACCEL', 'INFERENCE_BACKEND', 'LPR_CLASSES',
  ].concat(BOOL);

  const CHANNELS = [
    { id: 'telegram', label: 'Telegram' },
    { id: 'email', label: 'Email' },
    { id: 'pushover', label: 'Pushover' },
    { id: 'sms', label: 'SMS' },
    { id: 'webhook', label: 'Webhook' },
  ];

  let cameras = [];
  let rules = [];

  function toast(msg, ok) {
    const el = document.getElementById('int-toast');
    if (!el) return;
    el.textContent = msg;
    el.hidden = false;
    el.className = 'int-toast ' + (ok ? 'ok' : 'err');
    setTimeout(function () { el.hidden = true; }, 5000);
  }

  async function loadCameras() {
    const r = await fetch('/api/integrations/cameras');
    const j = await r.json();
    cameras = j.ok ? (j.cameras || []) : [];
  }

  async function loadRules() {
    const r = await fetch('/api/integrations/notification-rules');
    const j = await r.json();
    rules = j.ok ? (j.rules || []) : [];
    renderRules();
  }

  function camOptions(selected) {
    let html = '<option value="">All cameras</option>';
    cameras.forEach(function (c) {
      const sel = String(selected) === String(c.id) ? ' selected' : '';
      html += '<option value="' + c.id + '"' + sel + '>' + escapeHtml(c.name) + '</option>';
    });
    return html;
  }

  function escapeHtml(s) {
    return String(s || '').replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/"/g, '&quot;');
  }

  function channelChecks(rule, idx) {
    const active = rule.channels || [];
    return CHANNELS.map(function (ch) {
      const ck = active.indexOf(ch.id) >= 0 ? ' checked' : '';
      return '<label class="nr-ch"><input type="checkbox" data-idx="' + idx + '" data-ch="' + ch.id + '"' + ck + '>' + ch.label + '</label>';
    }).join('');
  }

  function renderRules() {
    const body = document.getElementById('nr-body');
    const empty = document.getElementById('nr-empty');
    if (!body) return;
    if (!rules.length) {
      body.innerHTML = '';
      if (empty) empty.hidden = false;
      return;
    }
    if (empty) empty.hidden = true;
    body.innerHTML = rules.map(function (rule, idx) {
      return (
        '<tr data-idx="' + idx + '">' +
        '<td><input type="checkbox" class="nr-enabled" data-idx="' + idx + '"' + (rule.enabled ? ' checked' : '') + '></td>' +
        '<td><select class="nr-camera" data-idx="' + idx + '">' + camOptions(rule.camera_id) + '</select></td>' +
        '<td><input type="text" class="nr-class" data-idx="' + idx + '" value="' + escapeHtml(rule.class_name || '*') + '" placeholder="person"></td>' +
        '<td><select class="nr-when" data-idx="' + idx + '">' +
        ['start', 'end', 'both', 'region'].map(function (w) {
          return '<option value="' + w + '"' + (rule.notify_on === w ? ' selected' : '') + '>' + w + '</option>';
        }).join('') + '</select></td>' +
        '<td><input type="number" class="nr-conf" data-idx="' + idx + '" min="0" max="1" step="0.05" value="' + (rule.min_confidence || 0) + '" style="width:4rem"></td>' +
        '<td><input type="number" class="nr-cd" data-idx="' + idx + '" min="5" max="3600" value="' + (rule.cooldown_s || 60) + '" style="width:4rem"></td>' +
        '<td class="nr-channels">' + channelChecks(rule, idx) + '</td>' +
        '<td><input type="checkbox" class="nr-snap" data-idx="' + idx + '"' + (rule.include_snapshot !== false ? ' checked' : '') + '></td>' +
        '<td><input type="checkbox" class="nr-clip" data-idx="' + idx + '"' + (rule.include_clip ? ' checked' : '') + '></td>' +
        '<td><button type="button" class="btn btn-secondary btn-sm nr-del" data-id="' + (rule.id || '') + '" data-idx="' + idx + '">×</button></td>' +
        '</tr>'
      );
    }).join('');

    body.querySelectorAll('.nr-del').forEach(function (btn) {
      btn.addEventListener('click', function () {
        deleteRule(parseInt(btn.getAttribute('data-id'), 10), parseInt(btn.getAttribute('data-idx'), 10));
      });
    });
    body.querySelectorAll('input, select').forEach(function (el) {
      el.addEventListener('change', function () { syncRuleFromRow(parseInt(el.getAttribute('data-idx'), 10)); });
    });
  }

  function syncRuleFromRow(idx) {
    const row = document.querySelector('tr[data-idx="' + idx + '"]');
    if (!row || !rules[idx]) return;
    const channels = [];
    row.querySelectorAll('.nr-ch input:checked').forEach(function (inp) {
      channels.push(inp.getAttribute('data-ch'));
    });
    const camSel = row.querySelector('.nr-camera');
    rules[idx] = {
      id: rules[idx].id,
      enabled: row.querySelector('.nr-enabled').checked,
      camera_id: camSel.value === '' ? null : parseInt(camSel.value, 10),
      class_name: row.querySelector('.nr-class').value.trim() || '*',
      notify_on: row.querySelector('.nr-when').value,
      min_confidence: parseFloat(row.querySelector('.nr-conf').value) || 0,
      cooldown_s: parseFloat(row.querySelector('.nr-cd').value) || 60,
      channels: channels,
      include_snapshot: row.querySelector('.nr-snap').checked,
      include_clip: row.querySelector('.nr-clip').checked,
    };
  }

  async function saveRule(rule) {
    const r = await fetch('/api/integrations/notification-rules', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(rule),
    });
    const j = await r.json();
    if (!j.ok) throw new Error(j.error || 'save rule failed');
    return j.rule;
  }

  async function saveAllRules() {
    for (let i = 0; i < rules.length; i++) {
      syncRuleFromRow(i);
      rules[i] = await saveRule(rules[i]);
    }
  }

  async function deleteRule(id, idx) {
    if (id) {
      await fetch('/api/integrations/notification-rules/' + id, { method: 'DELETE' });
    }
    rules.splice(idx, 1);
    renderRules();
  }

  document.getElementById('nr-add').addEventListener('click', function () {
    rules.push({
      enabled: true,
      camera_id: null,
      class_name: 'person',
      notify_on: 'start',
      min_confidence: 0.35,
      cooldown_s: 60,
      channels: ['telegram'],
      include_snapshot: true,
      include_clip: false,
    });
    renderRules();
  });

  async function loadSettings() {
    const r = await fetch('/api/integrations/settings');
    const j = await r.json();
    if (!j.ok) throw new Error(j.error);
    const d = j.data;
    FIELDS.forEach(function (k) {
      const el = document.getElementById(k);
      if (!el) return;
      if (el.type === 'checkbox') el.checked = !!d[k];
      else el.value = d[k] != null ? d[k] : '';
    });
  }

  function collectSettings() {
    const body = {};
    FIELDS.forEach(function (k) {
      const el = document.getElementById(k);
      if (!el) return;
      body[k] = el.type === 'checkbox' ? el.checked : el.value;
    });
    return body;
  }

  document.getElementById('int-form').addEventListener('submit', async function (e) {
    e.preventDefault();
    try {
      await saveAllRules();
      const r = await fetch('/api/integrations/settings', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(collectSettings()),
      });
      const j = await r.json();
      if (!j.ok) throw new Error(j.error);
      toast('Saved rules and settings.', true);
      await loadRules();
    } catch (err) {
      toast(err.message || String(err), false);
    }
  });

  document.getElementById('int-test-mqtt').addEventListener('click', async function () {
    try {
      const r = await fetch('/api/integrations/test-mqtt', { method: 'POST' });
      const j = await r.json();
      toast(j.message || j.error || 'Done', !!j.ok);
    } catch (err) { toast(String(err), false); }
  });

  document.querySelectorAll('.int-test').forEach(function (btn) {
    btn.addEventListener('click', async function () {
      try {
        const r = await fetch('/api/integrations/test-notify', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ channel: btn.getAttribute('data-channel') }),
        });
        const j = await r.json();
        toast(j.ok ? 'Test sent.' : (j.error || 'Failed'), !!j.ok);
      } catch (err) { toast(String(err), false); }
    });
  });

  (async function init() {
    try {
      await loadCameras();
      await loadSettings();
      await loadRules();
    } catch (err) {
      toast(err.message, false);
    }
  })();
})();

File: basebuddy/core/api/integrations.py
"""Integrations settings API (MQTT, notifications, recording modes)."""
from flask import Blueprint, jsonify, request

from basebuddy.core.config_persist import upsert_config_exports
from basebuddy.core.paths import get_repo_root

integrations_api = Blueprint("integrations_api", __name__)

_KEYS = [
    "MQTT_ENABLED", "MQTT_HOST", "MQTT_PORT", "MQTT_USERNAME", "MQTT_PASSWORD",
    "MQTT_TOPIC_PREFIX", "MQTT_CLIENT_ID",
    "NOTIFY_ENABLED", "NOTIFY_WEBHOOK_URL", "NOTIFY_PUBLIC_BASE_URL",
    "NOTIFY_FALLBACK_GLOBAL", "NOTIFY_INCLUDE_CLIP_DEFAULT",
    "TELEGRAM_BOT_TOKEN", "TELEGRAM_CHAT_ID",
    "PUSHOVER_USER_KEY", "PUSHOVER_API_TOKEN",
    "SMTP_HOST", "SMTP_PORT", "SMTP_USER", "SMTP_PASSWORD", "SMTP_FROM", "SMTP_TO", "SMTP_USE_TLS",
    "TWILIO_ACCOUNT_SID", "TWILIO_AUTH_TOKEN", "TWILIO_FROM_NUMBER", "TWILIO_TO_NUMBER",
    "RECORDING_MODE", "EVENT_CLIP_PRE_S", "EVENT_CLIP_POST_S",
    "FFMPEG_HWACCEL", "INFERENCE_BACKEND",
    "MULTIPROC_DETECTION", "LPR_ENABLED", "LPR_CLASSES",
]

_BOOL_KEYS = {
    "MQTT_ENABLED", "NOTIFY_ENABLED", "NOTIFY_FALLBACK_GLOBAL", "NOTIFY_INCLUDE_CLIP_DEFAULT",
    "MULTIPROC_DETECTION", "LPR_ENABLED", "SMTP_USE_TLS",
}


def _snapshot():
    from basebuddy.modules import config
    import basebuddy.modules.config as cfg_mod
    cfg_mod.load_config_file()
    out = {}
    for k in _KEYS:
        v = getattr(config, k, None)
        if isinstance(v, bool):
            out[k] = v
        elif v is None:
            out[k] = ""
        else:
            out[k] = v
    out["TELEGRAM_BOT_TOKEN_SET"] = bool(getattr(config, "TELEGRAM_BOT_TOKEN", ""))
    out["MQTT_PASSWORD_SET"] = bool(getattr(config, "MQTT_PASSWORD", ""))
    out["SMTP_PASSWORD_SET"] = bool(getattr(config, "SMTP_PASSWORD", ""))
    return out


@integrations_api.route("/api/integrations/settings", methods=["GET"])
def get_settings():
    try:
        return jsonify({"ok": True, "data": _snapshot()})
    except Exception as exc:
        return jsonify({"ok": False, "error": str(exc)}), 500


@integrations_api.route("/api/integrations/settings", methods=["POST"])
def save_settings():
    try:
        body = request.get_json(silent=True) or {}
        updates = {}
        for k in _KEYS:
            if k not in body:
                continue
            v = body[k]
            if k in _BOOL_KEYS:
                updates[k] = "true" if v in (True, "true", "1", 1, "on") else "false"
            else:
                updates[k] = str(v).strip() if v is not None else ""
        if updates:
            upsert_config_exports(get_repo_root(), updates)
            from basebuddy.modules.config import load_config_file
            load_config_file()
            if "MQTT_ENABLED" in updates:
                from basebuddy.core.services.mqtt_publisher import disconnect_mqtt
                disconnect_mqtt()
        return jsonify({"ok": True, "data": _snapshot()})
    except Exception as exc:
        return jsonify({"ok": False, "error": str(exc)}), 500


@integrations_api.route("/api/integrations/test-mqtt", methods=["POST"])
def test_mqtt():
    try:
        from basebuddy.core.services.mqtt_publisher import publish_event
        publish_event(0, "test", "new", {"message": "BaseBuddy MQTT test", "test": True})
        return jsonify({"ok": True, "message": "Test message published"})
    except Exception as exc:
        return jsonify({"ok": False, "error": str(exc)}), 500


@integrations_api.route("/api/integrations/test-notify", methods=["POST"])
def test_notify():
    try:
        body = request.get_json(silent=True) or {}
        channel = (body.get("channel") or "telegram").lower()
        from basebuddy.core.services.notification_service import send_test_notification
        result = send_test_notification(channel)
        return jsonify(result)
    except Exception as exc:
        return jsonify({"ok": False, "error": str(exc)}), 500


@integrations_api.route("/api/integrations/notification-rules", methods=["GET"])
def list_notification_rules():
    try:
        from basebuddy.core.services.notification_rules import list_rules
        return jsonify({"ok": True, "rules": list_rules()})
    except Exception as exc:
        return jsonify({"ok": False, "error": str(exc)}), 500


@integrations_api.route("/api/integrations/notification-rules", methods=["POST"])
def save_notification_rule():
    try:
        from basebuddy.core.services.notification_rules import save_rule
        body = request.get_json(silent=True) or {}
        rule = save_rule(body)
        return jsonify({"ok": True, "rule": rule})
    except Exception as exc:
        return jsonify({"ok": False, "error": str(exc)}), 500


@integrations_api.route("/api/integrations/notification-rules/<int:rule_id>", methods=["DELETE"])
def delete_notification_rule(rule_id):
    try:
        from basebuddy.core.services.notification_rules import delete_rule
        ok = delete_rule(rule_id)
        return jsonify({"ok": ok})
    except Exception as exc:
        return jsonify({"ok": False, "error": str(exc)}), 500


@integrations_api.route("/api/integrations/cameras", methods=["GET"])
def list_cameras_for_rules():
    try:
        import basebuddy.modules.state as st
        cams = []
        try:
            from basebuddy.modules.database import AnalyticsDB
            db = st.analytics_db or AnalyticsDB()
            with db._connect() as conn:
                cur = conn.execute(
                    "SELECT id, name FROM cameras WHERE enabled = 1 ORDER BY id"
                )
                for row in cur.fetchall():
                    cams.append({"id": row[0], "name": row[1] or f"Camera {row[0] + 1}"})
        except Exception:
            pass
        if not cams:
            for cid in sorted((st.grabbers or {}).keys()):
                cams.append({"id": cid, "name": f"Camera {cid + 1}"})
        return jsonify({"ok": True, "cameras": cams})
    except Exception as exc:
        return jsonify({"ok": False, "error": str(exc)}), 500
